lambda/family_separation.py

```diff
@@ -0,0 +1,198 @@
+from datetime import datetime, timedelta
+import io
+import json
+import logging
+import os
+import traceback
+from typing import Dict
+
+import boto3
+import requests
+from apiclient.discovery import build
+from mapbox import Geocoder
+from PIL import Image
+
+logging.basicConfig(level=logging.DEBUG)
+log = logging.getLogger(__name__)
+
+def read_sheet(sheet_range, fields, location_idx, affiliate):
+    log.info('\nload sheet %s with %s', os.environ['SHEET_ID'],
+             os.environ['GOOGLE_API_KEY'])
+    service = build('sheets', 'v4', developerKey=os.environ['GOOGLE_API_KEY'], cache_discovery=False)
+    result = service.spreadsheets().values().get(
+        spreadsheetId=os.environ['SHEET_ID'], range=sheet_range).execute()
+    values = result.get('values', [])
+    rows = {}
+    empty = {}
+    for field in fields:
+        empty[field] = ''
+    idx = 0
+    for row in values[1:]:
+        if not isinstance(location_idx, list):
+            location_idx = [location_idx]
+
+        if max(location_idx) >= len(row):
+            log.warning('Skipping row %d because no location', idx)
+            continue
+
+        props = {}
+        props.update(empty)
+        for field in fields:
+            idx = fields[field]
+            if idx < len(row):
+                val = row[idx].strip()
+                if val:
+                    props[field] = val
+
+                # Y|N to boolean
+                if props[field] == 'Y':
+                    props[field] = True
+                if props[field] == 'N':
+                    props[field] = False
+        location = ' '.join(map(lambda f: row[f].strip(), location_idx))
```

Comment:
Some rows in the spreadsheet have the same city/state/country combo, so it looks like this will cause them to have identical keys. The links should be unique, though, so maybe use that?